app/Http/Controllers/Ranking/DailyChallengeController.php

```diff
@@ -8,63 +8,76 @@
 namespace App\Http\Controllers\Ranking;
 
 use App\Http\Controllers\Controller;
+use App\Libraries\DailyChallengeDateHelper;
 use App\Models\Multiplayer\Room;
+use App\Models\Multiplayer\UserScoreAggregate;
 use Carbon\CarbonImmutable;
 use Carbon\Exceptions\InvalidFormatException;
 
 class DailyChallengeController extends Controller
 {
-    private const string DATE_FORMAT = 'Y-m-d';
-
-    public static function roomId(Room $room): string
-    {
-        return $room->starts_at->format(static::DATE_FORMAT);
-    }
-
     public function index()
     {
         $room = Room::dailyChallenges()->last() ?? abort(404);
 
-        return ujs_redirect(route('daily-challenge.show', ['daily_challenge' => static::roomId($room)]));
+        return ujs_redirect(route('daily-challenge.show', ['daily_challenge' => DailyChallengeDateHelper::roomId($room)]));
     }
 
     public function show(string $dateString)
     {
         try {
-            $date = CarbonImmutable::createFromFormat(static::DATE_FORMAT, $dateString);
+            $date = CarbonImmutable::createFromFormat(DailyChallengeDateHelper::DATE_FORMAT, $dateString)->startOfDay();
         } catch (InvalidFormatException) {
             abort(404, 'invalid date');
         }
 
-        $room = Room::dailyChallengeFor($date) ?? abort(404);
-        $playlist = $room->playlist[0];
+        $dateHelper = new DailyChallengeDateHelper($date);
 
-        $currentRoomOption = [
-            'id' => $dateString,
-            'text' => $dateString,
-        ];
-        $roomOptions = Room::dailyChallenges()
-            ->orderByDesc('id')
-            ->get()
-            ->map(static::roomId(...))
-            ->map(fn (string $roomName): array => [
-                'id' => $roomName,
-                'text' => $roomName,
-            ]);
+        $room = Room::dailyChallengeFor($date);
+        if ($room === null) {
+            $playlist = null;
+            $scores = null;
+            $userScore = null;
+        } else {
+            $playlist = $room->currentPlaylistItem;
 
-        $scores = $room->topScores()->paginate();
+            $scores = $room->topScores()->paginate();
```

Comment:
not this directly, but have we been doing the order without an index this whole time? 🤔